lib/Dialect/TTIR/Transforms/Quantization.cpp

```
@@ -0,0 +1,153 @@
+#include "ttmlir/Dialect/TTIR/Utils/UniformTypeRewriter.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Quant/IR/Quant.h"
+#include "mlir/Dialect/Quant/IR/QuantTypes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/Transforms/WalkPatternRewriteDriver.h"
+
+namespace mlir::tt::ttir {
+#define GEN_PASS_DEF_TTIRQUANTDATATYPECONVERSIONPASS
+#include "ttmlir/Dialect/TTIR/Transforms/Passes.h.inc"
+
+namespace {
+
+// Helper function to convert integer bit width string to integer type.
+static Type getIntegerTypeFromString(MLIRContext *context, StringRef bitWidth) {
+  // Default to int32 if empty.
+  if (bitWidth.empty()) {
+    return IntegerType::get(context, 32, IntegerType::Signed);
+  }
+
+  struct BitWidthMapping {
+    StringRef name;
+    unsigned width;
+  };
+
+  static const BitWidthMapping mappings[] = {
+      {"int8", 8}, {"int16", 16}, {"int32", 32}, {"int64", 64}};
+
+  for (const BitWidthMapping &mapping : mappings) {
+    if (bitWidth == mapping.name) {
+      return IntegerType::get(context, mapping.width, IntegerType::Signed);
+    }
+  }
+
+  return nullptr;
+}
+
+static std::pair<int64_t, int64_t> getStorageTypeMinMax(IntegerType intType) {
+  // Only need to return signed min and max.
+  int64_t min = -(1LL << (intType.getWidth() - 1));
+  int64_t max = (1LL << (intType.getWidth() - 1)) - 1;
+  return {min, max};
+}
+
+// Convert quantized type to use the target integer bit width.
+static Type convertQuantizedType(Type type, Type targetIntType) {
+  if (quant::QuantizedType quantType = dyn_cast<quant::QuantizedType>(type)) {
+    // Use the target integer type's min and max values.
+    IntegerType intType = cast<IntegerType>(targetIntType);
+    std::pair<int64_t, int64_t> storageTypeMinMax =
+        getStorageTypeMinMax(intType);
+    int64_t storageTypeMin = storageTypeMinMax.first;
+    int64_t storageTypeMax = storageTypeMinMax.second;
+
+    assert(intType.getWidth() >=
```

@nsmithtt @anuragsingh-tt Do you want to `assert` here? For element type normalization we didn't assert, just downcasted.

---

Maybe we should log an error and fail compile for the user instead of assert.

But I think this probably is an error, int quantization would probably just not work at all if you downcasted below the width that the model was trained on.  Maybe we could add a flag like `no-really-i-want-to-cast-it-trust-me`.

---

I'm not expert in the area but in ttmlir so far we downcasted to supported type. I decided to keep the same behaviour for ETN to not break FE models, but I think we should discuss this. Do you know which person from metal integrates models into their code base?

---

So the difference here is floating point vs integer quantized.  It's generally safe to demote floating point, you'll usually just get a degradation in precision/accuracy.  For quantized models though, they are very specifically trained and hand crafted for a specific integer bitwidth, if you demoted the int type you'd almost certainly break the model completely.

---

Yes, let's assert here. For quantized models, silently downcasting the integer type could break correctness in subtle ways, so it's better to fail fast and make the error explicit to the user.